jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -60,62 +51,61 @@ static ecma_number_t
 get_typedarray_element (lit_utf8_byte_t *src, /**< the location in the internal arraybuffer */
                         lit_magic_string_id_t class_id) /**< class name of the typedarray */
 {
-  ecma_number_t ret_num;
-
   switch (class_id)
   {
     case LIT_MAGIC_STRING_INT8_ARRAY_UL:
     {
-      GET_ELEMENT (int8_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int8_t, src);
     }
     case LIT_MAGIC_STRING_UINT8_ARRAY_UL:
     case LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL:
     {
-      GET_ELEMENT (uint8_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint8_t, src);
     }
     case LIT_MAGIC_STRING_INT16_ARRAY_UL:
     {
-      GET_ELEMENT (int16_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int16_t, src);
     }
     case LIT_MAGIC_STRING_UINT16_ARRAY_UL:
     {
-      GET_ELEMENT (uint16_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint16_t, src);
     }
     case LIT_MAGIC_STRING_INT32_ARRAY_UL:
     {
-      GET_ELEMENT (int32_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int32_t, src);
     }
     case LIT_MAGIC_STRING_UINT32_ARRAY_UL:
     {
-      GET_ELEMENT (uint32_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint32_t, src);
     }
     case LIT_MAGIC_STRING_FLOAT32_ARRAY_UL:
     {
-      GET_ELEMENT (float, src, ret_num);
-      break;
+      return GET_ELEMENT (float, src);
     }
 #if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
     case LIT_MAGIC_STRING_FLOAT64_ARRAY_UL:
     {
-      GET_ELEMENT (double, src, ret_num);
-      break;
+      return GET_ELEMENT (double, src);
     }
 #endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
     default:
     {
       JERRY_UNREACHABLE ();
+      return 0;
     }
   }
-
-  return ret_num;
 } /* get_typedarray_element */
 
+#undef GET_ELEMENT
+
+/**
+ * Cast the number from ecma_numbet_t into type,
+ * and store the number in the location
+ */
+#define SET_ELEMENT(type, location, number) \
+  int32_t v = (int32_t) number; \
+  *((type *) location) = (type) v;
```

Comment:
castiong to int32_t is not sufficient. See the test case 22.02.01-015, 017, 018, 019.

we are casting a float number (the ecma_number_t) to interger, we should make sure the integral part of float number can be represented by the integer type. So I choose int64_t to make sure that.
```
When a finite value of real floating type is converted to an integer type other than _Bool, the fractional part is discarded (i.e., the value is truncated toward zero). If the value of the integral part cannot be represented by the integer type, the behavior is undefined.
```

http://c0x.coding-guidelines.com/6.3.1.4.html